url9d1719148e_mrqsdavi_pegmachine_tgz-pJ8-peg_RegexJ8/src/peg/Regex.java

```diff
@@ -0,0 +1,616 @@
+				previous = instrucoes.get(0);
+				for(int i = 0; i < instrucoes.size(); i++){
+					if(instrucoes.get(i).getInstrucaoDesvio() == previousTemporaria){
+						instrucoes.get(i).setInstrucaoDesvio(previous);
+					}
+				}
+				
+				retorno.addAll(instrucoes);
+			}
+			
+			if(useSpanOptimization && padrao.repeticao().getPadrao().getTipo() == TipoPadrao.CONJUNTO){
+				
+				retorno.add(new ISpan(padrao.repeticao().getPadrao().conjunto().getConjuntoCaracteres()));
+				
+			}else{
+				
+				IChoice choiceRepeticao = new IChoice("");
+				ICommit commitRepeticao = new ICommit("");
+				
+				choiceRepeticao.setInstrucaoDesvio(next);
+				commitRepeticao.setInstrucaoDesvio(choiceRepeticao);
+				
+				previous = choiceRepeticao;
+				
+				ArrayList<Instrucao> instrucoesRepeticao = instrucoesDoPadrao(padrao.repeticao().getPadrao());
+				
+				retorno.add(choiceRepeticao);
+				retorno.addAll(instrucoesRepeticao);
+				
+				if(usePartialCommitOptimization){
+					for(int i = 0; i < instrucoesRepeticao.size(); i++){
+						if(instrucoesRepeticao.get(i).getInstrucaoDesvio() == previous){
+							instrucoesRepeticao.get(i).setInstrucaoDesvio(instrucoesRepeticao.get(0));
+						}
+					}
+					
+					IPartialCommit partialCommit = new IPartialCommit("");
+					partialCommit.setInstrucaoDesvio(instrucoesRepeticao.get(0));
+					retorno.add(partialCommit);
+					
+				}else{
+					retorno.add(commitRepeticao);
+				}
+				
+			}
+			
+		}
+		break;
+		
+		case CONCATENACAO:{
+			
+			ArrayList<Instrucao> instrucoesTemporarias = new ArrayList<>();
+			ArrayList<Instrucao> instrucoesReais = new ArrayList<>();
+			
+			for(int i = 0; i<padrao.concatenacao().getPadroes().size();i++){
+				instrucoesTemporarias.add(new ICall("L"+i));
+			}
+			
+			for(int i = 0; i<padrao.concatenacao().getPadroes().size();i++){
+				
+				Padrao padraoAtual = padrao.concatenacao().getPadroes().get(i);
+				Instrucao instrucaoPrevious = null;
+				Instrucao instrucaoNext = null;
+				
+				if(i==0){
+					instrucaoPrevious = backupPrevious;
+				}else{
+					instrucaoPrevious = instrucoesTemporarias.get(i-1);
+				}
+				
+				if(i==padrao.concatenacao().getPadroes().size()-1){
+					instrucaoNext = backupNext;
+				}else{
+					instrucaoNext = instrucoesTemporarias.get(i+1);
+				}
+				
+				previous = instrucaoPrevious;
+				next = instrucaoNext;
+
+				ArrayList<Instrucao> instrucoesPadraoAtual = instrucoesDoPadrao(padraoAtual);
+				instrucoesReais.add(instrucoesPadraoAtual.get(0));
+				retorno.addAll(instrucoesPadraoAtual);
+				
+			}
+			
+			for(int i = 0; i < retorno.size(); i++){
+				Instrucao instrucaoAtual = retorno.get(i);
+				if(instrucaoAtual.getInstrucaoDesvio()!=null && instrucoesTemporarias.contains(instrucaoAtual.getInstrucaoDesvio())){
+					int index = instrucoesTemporarias.indexOf(instrucaoAtual.getInstrucaoDesvio());
+					retorno.get(i).setInstrucaoDesvio(instrucoesReais.get(index));
+				}
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 278:284, 354:360.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.